File: Easy/387. First Unique Character in a String/solution.java
class Solution {
  public int firstUniqChar(String s) {
      Map<String, Integer> dic = new HashMap<>();
      int ans = 999999;

      for (int i=0; i<s.length(); i++){
          if (dic.containsKey(String.valueOf(s.charAt(i)))) {
              dic.put(String.valueOf(s.charAt(i)), -1);
          } else {
              dic.put(String.valueOf(s.charAt(i)), i);
          }
      }
      int max_v = -100;
      for (Map.Entry<String, Integer> entry : dic.entrySet()) {
          if (max_v < entry.getValue()) {
              max_v = entry.getValue();
          }
      }
      if (max_v == -1) {
          return -1;
      } else {
          for (Map.Entry<String, Integer> entry : dic.entrySet()) {
              if (entry.getValue() == -1){
                  continue;
              } else if (ans > entry.getValue()) {
                  ans = entry.getValue();
              }
          }
          return ans;
      }
  }
}

// Time Complexity: O(n)
// Space Complexity: O(n)

// other solution
class Solution {
  public int firstUniqChar(String s) {
      HashMap<Character, Integer> mp = new HashMap<>();

      for (char a : s.toCharArray()) {
          mp.put(a, mp.getOrDefault(a, 0) + 1);
      }

      for (int i = 0; i < s.length(); i++) {
          if (mp.get(s.charAt(i)) == 1) {
              return i;
          }
      }
      return -1;
  }
}

File: Easy/387. First Unique Character in a String/solution.py
class Solution(object):
    def firstUniqChar(self, s):
        """
        :type s: str
        :rtype: int
        """
        dic = {}
        ans = 999999
        for i, c in enumerate(s):
            if c in dic:
                dic[c] = -1
            else:
                dic[c] = i
        if max(dic.values()) == -1:
            return -1
        else:
            for n in dic.values():
                if n == -1:
                    continue
                if ans > n:
                    ans = n
            return ans

# Time complexity: O(n)
# Space complexity: O(n)

# another solution
import collections
class Solution(object):
    def firstUniqChar(self, s):
        hset = collections.Counter(s)
        # Traverse the string from the beginning...
        for idx in range(len(s)):
            # If the count is equal to 1, it is the first distinct character in the list.
            if hset[s[idx]] == 1:
                return idx
        return -1       # If no character appeared exactly once...